crates/vm/levm/src/hooks/l2_hook.rs

```diff
@@ -142,81 +70,104 @@ impl Hook for L2Hook {
-                    TxValidationError::InsufficientMaxFeePerGas.into(),
-                );
-                ctx_result.gas_used = vm.current_call_frame.gas_limit;
-                ctx_result.output = Bytes::new();
-                // Set l1_gas to use all remaining gas
-                l1_gas = vm
-                    .current_call_frame
-                    .gas_limit
-                    .saturating_sub(actual_gas_used);
+    let mut l1_gas = calculate_l1_fee_gas(
+        vm,
+        std::mem::take(pre_execution_backup),
+        &fee_config.l1_fee_config,
+    )?;
 
-                total_gas = vm.current_call_frame.gas_limit
-            }
+    let mut total_gas = actual_gas_used
+        .checked_add(l1_gas)
+        .ok_or(InternalError::Overflow)?;
 
-            delete_self_destruct_accounts(vm)?;
+    if total_gas > vm.current_call_frame.gas_limit {
+        vm.substate.revert_backup();
+        vm.restore_cache_state()?;
 
-            // L1 fee is paid to the L1 fee vault
-            pay_to_l1_fee_vault(vm, l1_gas, self.fee_config.l1_fee_config)?;
+        undo_value_transfer(vm)?;
 
-            refund_sender(vm, ctx_result, gas_refunded, total_gas)?;
+        ctx_result.result =
+            crate::errors::TxResult::Revert(TxValidationError::InsufficientMaxFeePerGas.into());
+        ctx_result.gas_used = vm.current_call_frame.gas_limit;
+        ctx_result.output = Bytes::new();
 
-            // We pay to coinbase after the l1_fee to avoid charging the diff to every transaction
-            pay_coinbase_l2(vm, actual_gas_used, &self.fee_config.operator_fee_config)?;
+        l1_gas = vm
+            .current_call_frame
+            .gas_limit
+            .saturating_sub(actual_gas_used);
+        total_gas = vm.current_call_frame.gas_limit;
+    }
 
-            // Base fee is not burned
-            pay_base_fee_vault(vm, actual_gas_used, self.fee_config.base_fee_vault)?;
+    delete_self_destruct_accounts(vm)?;
 
-            // Operator fee is paid to the chain operator
-            pay_operator_fee(vm, actual_gas_used, &self.fee_config.operator_fee_config)?;
+    if let Some(l1_fee_config) = fee_config.l1_fee_config {
+        pay_to_l1_fee_vault(vm, l1_gas, l1_fee_config, use_fee_token)?;
+    }
 
-            ctx_result.gas_used = total_gas;
+    if use_fee_token {
+        refund_sender_fee_token(vm, ctx_result, gas_refunded, total_gas)?;
+    } else {
+        refund_sender(vm, ctx_result, gas_refunded, total_gas)?;
+    }
 
-            return Ok(());
-        }
+    pay_coinbase_l2(
+        vm,
+        actual_gas_used,
+        &fee_config.operator_fee_config,
+        use_fee_token,
+    )?;
+
+    if let Some(base_fee_vault) = fee_config.base_fee_vault {
+        pay_base_fee_vault(vm, actual_gas_used, base_fee_vault, use_fee_token)?;
+    } else if use_fee_token {
+        pay_base_fee_vault(vm, actual_gas_used, Address::zero(), use_fee_token)?;
+    }
```

Comment:
Nit. (it isn't formatted).
```suggestion
    pay_base_fee_vault(vm, actual_gas_used, fee_config.base_fee_vault.unwrap_or_default(), use_fee_token)?;
```